### PR TITLE
Jackie & Cassy - OO Rideshare - Nodes

### PR DESCRIPTION
**Pull request by jacquelynoelle**
# OO Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | One inheritance design decision we made was to call the super net_expenditures method in User from the Driver method. Originally, we had written the Driver method totally independently, then realized the return value from the super method could replace some of our code. |
Describe any examples of composition that you encountered in this project, if any | After writing our tests, we realized our code was very not DRY, and we made the decision to take a new look at composition of our test classes. We decided to refactor and put a substantial amount of code in before do blocks in our describes (both at the class level and test group level). |
Describe the relationship between `User` and `Driver`| User is the super class for Driver. Driver is a User. Driver inherits variables like id, name, etc. from User, as well as some methods like net_expenditures. |
Describe a _nominal test_ that you wrote for this assignment. | We checked to make sure methods like total_revenue and net_expenditures were correctly calculating their return values. |
Describe an _edge case test_ that you wrote for this assignment | For each of our methods that calculated totals, we wrote tests to ensure in progress trips would not be erroneously included. |
Describe a concept that you/your pair gained more clarity on as you worked on this assignment | Debugging tests! We also learned a lot about using pry to debug, and using git in a ping pong style to test and code. |
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | 1. Before starting, Cassy mentioned she preferred if her partner could wait to provide feedback/input until there is a pause in coding to let her finish her thoughts. 2. In between waves or sections of waves we talked about who wanted to do tests vs methods for that piece of the project. 3. We both felt we did a good job of debugging, refactoring, and streamlining our code. 4. Jackie thought Cassy did a great job of adding comments to code for clarity. 5. Overall, great experience! |

### CODE REVIEW

File: specs/user_spec.rb [hunk 1]
```diff
@@ -59,4 +70,31 @@
...
+      expect(@no_trips_user.net_expenditures).must_equal 0
+    end
+  end
+
+  describe "total_time_spent method" do
+    it "calculates the total amount of time spent for all trips and does not include in-progress trips" do
+      # Adding in-progress trip
+      @user.add_trip(@inprogress_trip)
+      # Total should not include in progress trip
+      expect(@user.total_time_spent).must_equal (@user.trips[0].duration +
```
Comment (dHelmgren) [inline specs/user_spec.rb:92]: In my mind, asking the `user` about its trips is fraught. I'd keep this test consistent with your others, where you refer to the trips you made in the `before` block.

File: lib/driver.rb [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    end
+
+    def total_revenue
+      return 0.0 if driven_trips.empty?
+      completed = filter_completed_trips(driven_trips)
+      return ((completed.sum {|completed_trip| completed_trip.cost - 1.65}) * 0.8).round(2)
+    end
+
+    def net_expenditures
+      completed_riden = filter_completed_trips(trips)
```
Comment (dHelmgren) [inline lib/driver.rb:51]: Is this line necessary?

File: lib/trip_dispatcher.rb [hunk 3]
```diff
@@ -67,10 +61,119 @@ def inspect
...
+        selected_driver = available_drivers_no_trips.first
+      else
+        # Pick the next available driver who has not driven in the longest time
+        available_drivers.sort_by! { |driver|
+          driver.driven_trips.max_by { |trip|
+            trip.end_time
+          }.end_time
+        }
+        selected_driver = available_drivers.first
+      end
```
Comment (dHelmgren) [inline lib/trip_dispatcher.rb:159]: The process here is somewhat convoluted to outside eyes, and really, if this big `if` statement is either going to return a driver or raise an error, it makes sense to turn it into a helper method if only to help readers ignore the nitty gritty.

**Comment by dHelmgren** [comment]
# Ride Share
## What We're Looking For

Feature|Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | A HEALTHY NUMBER OF COMMITS!
Answer comprehension questions | Hey, I want to call this out and clarify: When we talk about "composition" we mean "what one-to-one and one-to-many relationships exist". Composition means "class x is composed of class y" in some form.
**Wave 1** |
Appropriate use of Ruby's `Date`| Yep!
`Trip` has a helper method to calculate duration| Yes!
`User` (passenger) has a method to calculate total cost of all trips| yep!
Tests for wave 1 | Looking good!
**Wave 2** |
`Driver` inherits from `User` | Check!
`Driver` has `add_driven_trip` method | check
`Driver` has method to calculate average rating | check! 
`Driver` has method to calculate net expenditures and it uses `super` | I think you left a debug line in? Check!
`Driver` has a method to calculate total revenue | Check!
Tests for wave 2 | Nice work!
**Wave 3** |
`TripDispatcher` has a new method to create trips | Check!
creating a trip in `TripDispatcher` relies on methods in `Driver` and `User` (passenger) to modify their own attributes | Check!
Complex logic was correctly implemented | Yes!
Tests for `request_trip` | They're REAL!
Methods from wave 1 and 2 handle incomplete trips | They do!
Tests for wave 1 and 2 methods with incomplete trips | YEAH!
**Wave 4 (Optional)** |
`TripDispatcher` now assigns trips to either the newest driver (no trips), or the driver who has not driven in the longest time | Great work!
Appropriate helper methods were made to help with complex logic | You might have been able to break your method into smaller pieces. see my comments.
Tests for wave 4 | Solid tests!
**Overall** | This is a really great submission from the both of you! Nice work across the board.
